contracts/src/PDPService.sol

```diff
@@ -3,15 +3,151 @@ pragma solidity ^0.8.13;
+        uint256[] sizes;
+        uint256[] sumTree;
+        uint256 size;
+        address owner;
+        nextRootID uint64;
+    }
+    ** PDP service contract tracks many possible proof sets **
+    []ProofSet proofsets
+
+    To implement this logical structure in the solidity data model we have
+    two arrays tracking the singleton fields and three two dimensional arrays
+    tracking the growing data of the proof set.  The first index is the proof set id
+    and the second index is the index of the data in the array.
+
+    Invariant: rootCids.length == rootSizes.length == sumTreeSizes.length
+    */
+
+    // Network epoch delay between last proof of possession and next 
+    // randomness sampling for challenge generation
+    uint256 challengeFinality;
+   
+    // TODO PERF: https://github.com/FILCAT/pdp/issues/16#issuecomment-2329838769
+    uint64 nextProofSetId;
+    mapping(uint256 => mapping(uint256 => Cid)) rootCids; 
+    mapping(uint256 => mapping(uint256 => uint256)) rootSizes;
+    mapping(uint256 => mapping(uint256 => uint256)) sumTreeSizes;
+    mapping(uint256 => uint256) nextRootId;
+    mapping(uint256 => uint256) proofSetSize;
+    // ownership of proof set is initialized upon creation to create message sender 
+    // proofset owner has exclusive permission to add and remove roots and delete the proof set
+    mapping(uint256 => address) proofSetOwner;
 
     // Methods
-    constructor(uint256 _challengeFinality) Ownable(msg.sender) {
+    constructor(uint256 _challengeFinality) {
         challengeFinality = _challengeFinality;
     }
 
-}
\ No newline at end of file
+    // Returns the current challenge finality value
+    function getChallengeFinality() public view returns (uint256) {
+        return challengeFinality;
+    }
+
+    // Returns the next proof set ID
+    function getNextProofSetId() public view returns (uint64) {
+        return nextProofSetId;
+    }
+
+    // Returns the size of a proof set
+    function getProofSetSize(uint256 setId) public view returns (uint256) {
+        require(setId < nextProofSetId, "Proof set ID out of bounds");
+        return proofSetSize[setId];
+    }
+
+    // Returns the next root ID for a proof set
+    function getNextRootId(uint256 setId) public view returns (uint256) {
+        require(setId < nextProofSetId, "Proof set ID out of bounds");
+        return nextRootId[setId];
+    }
+
+    // Returns the owner of a proof set
+    function getProofSetOwner(uint256 setId) public view returns (address) {
+        require(setId < nextProofSetId, "Proof set ID out of bounds");
+        return proofSetOwner[setId];
+    }
+
+    // Returns the root CID for a given proof set and root ID
+    function getRootCid(uint256 setId, uint256 rootId) public view returns (Cid memory) {
+        require(setId < nextProofSetId, "Proof set ID out of bounds");
+        require(rootId < nextRootId[setId], "Root ID out of bounds");
+        return rootCids[setId][rootId];
```

Comment:
Here and in all the lookups above, this returns a default value (empty CID?) if the key is absent. This could be fine semantics, though might deserve some thought and documentation. We'll need to be careful of places where zero is an allowed value. For example, should we set nextProofSetID to 1 at construction, and similarly start root IDs at 1 so that zero can distinguish "no such proof set"?

If we do go with returning zero values for undefined mappings, there is not much point requiring the proof set ID or root ID to be within bounds. Is there any point distinguishing between not-yet-created vs deleted proof sets.

Ok to defer this, but it might affect multiple methods.

Reply:
As far as I can tell one big downside with no IPLD is that we can't delete a whole map in constant time so we're stuck with all the leftover root IDs after "deleting" a proof set.  This is going to give us some weird results without some care.

Reply:
Yeah with more research it looks like the only way to actually delete state is to iterate all map keys which is possible but so expensive that no-one would do it.

If we want to support logical proof set remove we need to handle queries that 
1) overrun the proof set index
2) overrun the root id index
3) hit deleted proof sets
4) hit deleted root ids 

I'd like to do this consistently.  The two possibilities are revert or return zero value.  I vaguely prefer return zero value: ints 0, cids zero length byte arrays.

To do this correctly we need to check for all conditions as part of the view methods and explicitly return zero values in some cases.

Reply:
After a few minutes of implementing I'm now prefering revert since there are so many view functions and the zero return value makes different amounts of sense for different functions

Reply:
Ok i've settled on consistently reverting queries to deleted or not created proof sets with "Proof set not live" but returning zero values for the proof set arrays (size, sumtree, roots).  SumTree needs to do this for deleted roots and zero values are less confusing when you don't need to figure out whether its proof set or array that isn't live.